Add Cavium SoC support

PatrickRudolph opened the pull request:
Add Cavium SoC support based on Cavium's latest ATF fork / CAVIUM-OCTEAN-TX-SDK-6.2.
Tested on Cavium CN81XX SFF EVB using coreboot as firmware.

antonio-nino-diaz-arm commented:
I haven't taken a look at the code, this is unmanageable. This pull request is bigger than the whole Trusted Firmware codebase. You need to split this into smaller chunks somehow. It's not like we can review the SATA driver, for example, but it's still pretty hard to do anything with this PR.

However, there is a bigger problem. Some files don't have a copyright header, is this because you forgot to add it? And, in the files that there is a header, they don't have the standard header of the rest of the codebase.

PatrickRudolph commented:
I've added the SPDX license header to all files, removed some headers and cleaned the remaining ones.
I've split the some commits into smaller ones.
Please review.

antonio-nino-diaz-arm commented:
Ok, it certainly looks a lot better now. You don't need to split the stdlib patches into different ones, by the way. The comment was about the rest of the patches.

We'll take a look at it, but there are still some copyright headers missing (for example ``include/drivers/hw_timers.h``) and many others that don't use the SPDX identifier.

Also, you'll want to add your platform to the maintainers.rst and maybe Cavium to the acknowledgements.rst file.

antonio-nino-diaz-arm commented:
I'm taking a look at this. The commit "cavium: Add sata quirks is the only one that is missing some copyrights.

Regarding the commit "[Quirk]cavium: GPIO IRQ to EL0", you are doing a lot of changes to common context management code, and the commit message itself doesn't clarify the reason why:

> Includes lots of hacks and it's unclear if useful. Cavium does not provide a sample code and neither explains why intercepting interrupts could be useful. This feature isn't used on any reference board. Only included for reference.

If there is no clear reason to have this commit, it should be dropped, as it affects other platforms as well.

PatrickRudolph commented:
Dropped the GPIO IRQ to EL0 commit.
Split the other commits into smaller chunks.
Rebased on latest master.
Fixed more license headers.
Moved timers code from /lib to /drivers/timers and implemented a clear API.
Added Cavium to acknowledgements.rst file

I'm not going to add me to maintainers, as I don't get paid for it and I'm busy in my free time.

danh-arm commented on docs/plat/cavium/atf_abi.txt:
```diff
@@ -0,0 +1,139 @@
+CAVIUM
```
None of these contributed docs explain how to use the platform firmware. I would expect a separate document that explains as a minimum:

*   What the platform is and where to get more information.

*   How to build the platform firmware, including configuration options and any pre-requisites.

*   How to install the firmware (or links to this).

PatrickRudolph replied:
Will add it.

danh-arm commented on acknowledgements.rst:
```diff
@@ -14,5 +14,7 @@ Xilinx, Inc.
 NXP Semiconductors
 
+Cavium, Inc.
+
```
Have you been authorized by Cavium to make this contribution on their behalf?

PatrickRudolph replied:
I don't understand. Does the acknowledgements contain a list of contributors or a list of companies who developed the submitted code ?

zaolin replied:
@danh-arm The firmware was open source by cavium on github and released under BSD license. So it should be fine

danh-arm replied:
This file should contain the companies that contributed if the contributor is an employee of that company, or the individual names of contributors that do not work for a particular company. In your case, it sounds like you are trying to upstream code from another GitHub project without Cavium's consent or knowledge. If that is the case, then it is your company (9Elements) that should be listed here, not Cavium. Not only that but we need to assess the provenance of that code and the value of it being in the upstream. If no-one is prepared to maintain this code, then it might as well remain out of tree. Can you provide a link to the source GitHub project?

rchintakuntla replied:
@danh-arm and @PatrickRudolph I am from Cavium and we are not aware of this upstreaming support of OcteonTx. And I have confirmed that no one from Cavium OcteonTx group has authorized this activity. Since the original authors of this whole code has been modified it will not be Cavium's responsiblity to fix any bugs in this code and maintain it in future. 
Also we have chip errata that is not public and so 9elements could land in trouble. I have forwarded this to our legal team and they will take appropriate action.

zaolin replied:
We never modified any license headers. All rights belong to Cavium Inc. We just upstreamed the already public available code under BSD license

dhendrix replied:
@rchintakuntla FYI, the code is from the Octeon-TX SDK release here: https://github.com/Cavium-Open-Source-Distributions/OCTEON-TX-SDK

It will be mutually beneficial for Cavium and their customers to have ATF support upstreamed. I'd like to discuss this further with management to see if we can get resources devoted to this effort. Working with the unsupported, out-of-tree ATF from the tarball on Github is sub-optimal for everybody (especially on our end).

rchintakuntla replied:
@dhendrix and @zaolin  having code on github doesn't authorize anyone to upstream code. Since ATF is opensource we publish it on github along with lot other stuff. It is purely a FB and 9Elements thing now and Cavium has no role in this. Please remove Cavium Inc from maintainers atleast so that we will not be held responsible for maintaining. This is old piece of code anyway.
I am not sure what will fB gain from upstreaming this. None of our customers are going to benefit from this as they modify a lot on what we provide.

dhendrix replied:
@rchintakuntla I'll set up a meeting with my contacts at Cavium to discuss this further.

It is unexpected to run into this kind of resistance to upstreaming open-source code to support Cavium products. Those present in previous meetings seemed onboard and Cavium is active in upstream Linux kernel development. But if Cavium doesn't want to participate in upstream ATF development then we'll just work off a fork like everyone else; I don't see any technical benefits to that approach but it is what it is.

rchintakuntla replied:
@dhendrix you are misinterpreting my statements. What I meant is: if FB wants to upstream Cavium's old ATF code for whatever reason they seemed fit they can go ahead but please DO NOT PUT Cavium as maintainer as it totally seems pointless. It's an old code and I do not see a benefit. Also the authorship has been hijacked which I am totally against. Put maintainer as either facebook or 9elements then Cavium can be off the hook. This is the only reason I responded as @danh-arm pointed this as well. I don't know why you think this is resistance. Maybe you are new to upstreaming ? 

Had it been FB requesting Cavium actual patchset against current ATF code base then that would be a meaningful approach and then can expect some participation from Cavium.

BTW what Cavium upstreams and what it doesn't depends on Cavium's business requirements and resources to do that. For one example, if we upstream our ATF I do not see a point where some individual associated with Cavium as a customer is going to benefit. There are lot of pieces missing plus we do lot of custom modifications to satisfy some customer requirements. Now upstreaming these will be a pain and might break other platforms. Linux kernel is a different ball game and there's a very good reason for that. I am sure you know that.

danh-arm commented on drivers/cavium/thunder_mmc.c:
```diff
@@ -0,0 +1,425 @@
+/*
+ * Copyright (c) 2016-2017, Cavium Inc. All rights reserved.<BR>
```
Please use the standard copyright that the rest of the codebase uses (Arm Limited and Contributors), assuming you are authorised by Cavium to make this contribution on their behalf.

PatrickRudolph replied:
No I can't change the license.

danh-arm replied:
I was talking the copyright notice, not the license. But you're right that you can't change this if you are not working on behalf of Cavium

danh-arm commented on drivers/cavium/thunder_mmc.c:
```diff
@@ -0,0 +1,425 @@
+/*
```
Have Cavium authorised the release of open source drivers for the hardware IP in this contribution, in particular the eMMC, SPI, SATA and TWSI drivers? Are there publicly available data sheets for these IPs?

PatrickRudolph replied:
Yes the whole code was release under BSD license.
There's no data sheet publicly available.

danh-arm replied:
Released by Cavium? Please provide a link to the source tree

PatrickRudolph replied:
It's available here. You have to use git LFS and then install the RPM...
https://github.com/Cavium-Open-Source-Distributions/OCTEON-TX-SDK

danh-arm commented on lib/stdlib/strncmp.c:
```diff
@@ -50,3 +51,22 @@ strncmp(const char *s1, const char *s2, size_t n)
+		us2++;
+	} while (--n != 0);
+	return 0;
+}
```
Is this code your own work? In which case, please update the year in the "Portions copyright" notice above. If not please, state the original source of the code in the commit message, correctly attributing copyright where needed.

danh-arm commented on lib/stdlib/strncpy.c:
```diff
@@ -0,0 +1,57 @@
+/*	$OpenBSD: strncpy.c,v 1.7 2014/06/10 04:17:37 deraadt Exp $	*/
```
Please state in the commit message the precise repository location where you are importing this code from. All the other stdlib code in TF-A is from FreeBSD so it's preferable to use that repo.

danh-arm commented on plat/cavium/common/bl31_thunder_setup.c:
```diff
@@ -0,0 +1,319 @@
+/*
+ * Copyright (c) 2013-2014, ARM Limited and Contributors. All rights reserved.
+ * Copyright (c) 2016-2017, Cavium Inc. All rights reserved.<BR>
+ * Copyright (c) 2018     , Facebook, Inc.
```
Are you authorised to contribute this code on behalf of Facebook (and Cavium)? If so, please use only the first copyright notice.

PatrickRudolph replied:
Yes.
Though changing the copyright seems tricky.
IANAL but it seems wrong, even when it's released under BSD license.

From BSD license:
"a. Redistributions of source code must retain the above copyright notice,"

danh-arm replied:
If you are authorised to contribute this code on behalf of Facebook and Cavium then you can ask their permission to change the copyright notice. It would not change who the underlying copyright holders are. But it sounds like you don't have such authorisation from earlier comments.

dhendrix replied:
Hello, I'm an engineer from Facebook working with 9elements and Cavium. From our standpoint, yes, 9elements has authorization to release this code. Since 9elements is working for Facebook on contract, the copyright change made here is appropriate AFAICT.

Some additional background: The sources here are derived from Cavium's Octeon-TX SDK: https://github.com/Cavium-Open-Source-Distributions/OCTEON-TX-SDK

Basically you'll find a ~1 year old copy of ATF in there, which we've forward ported to upstream. 9elements has driven mostly driven this effort and has tested it.

hzhuang1 commented on include/drivers/io/io_spi.h:
```diff
@@ -0,0 +1,14 @@
+
+int register_io_dev_spi(const struct io_dev_connector **dev_con);
+
+#endif /* __IO_SPI_H__ */
```
Move this into another commit. Don't modify common file and platform driver in the same commit.

hzhuang1 commented on include/drivers/io/io_storage.h:
```diff
@@ -22,6 +22,8 @@ typedef enum {
 	IO_TYPE_FIRMWARE_IMAGE_PACKAGE,
 	IO_TYPE_BLOCK,
+	IO_TYPE_SPI,
+	IO_TYPE_EMMC,
```
Do we need to add a new IO_TYPE_EMMC? I think IO_TYPE_BLOCK is good to eMMC device.

danh-arm commented:
I'm closing this PR for now. Another one can be raised once the IP and ownership issues have been resolved.
